framework/src/problems/FEProblemBase.C

```diff
@@ -6324,7 +6324,18 @@ FEProblemBase::meshChangedHelper(bool intermediate_change)
                                     _material_props,
                                     _bnd_material_props,
                                     _assembly);
-      Threads::parallel_reduce(*_mesh.refinedElementRange(), pmp);
+      const auto & range = *_mesh.refinedElementRange();
+      Threads::parallel_reduce(range, pmp);
+
+      // Concurrent erasure from the shared hash map is not safe while we are reading from it in
+      // ProjectMaterialProperties, so we handle erasure here. Moreover, erasure based on key is not
+      // thread safe in and of itself because it is a read-write operation
+      for (auto it = range.begin(); it != range.end(); ++it)
+      {
+        const Elem * const elem = *it;
```

Comment:
```suggestion
      for (const auto & elem : make_range(range.begin(), range.end())
      {
```

for future reference

Reply:
?! why make a range when you already have a range? couldn't you just do

```
      for (const auto & elem : range)
```

Reply:
oops